File: crypto/src/test/cplusplus/fermat.cpp
/*
 * Copyright (c) 2024 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <boost/test/unit_test.hpp>

#include "fermat.h"

using Zq = FermatRing;

BOOST_AUTO_TEST_SUITE(Fermats)

BOOST_AUTO_TEST_CASE(ZqAdd) {
    Zq a(981);
    Zq b(-1516);
    Zq c(-535);
    BOOST_TEST(c == a + b);
    BOOST_TEST(c == b + a);
    BOOST_TEST(c == Zq(0) + c);
    BOOST_TEST(c == c + Zq(0));
    BOOST_TEST(Zq(1) == Zq(1) + Zq(0));
    BOOST_TEST(Zq(1) == Zq(0) + Zq(1));
    BOOST_TEST(Zq(0) == Zq(-1) + Zq(1));
}

BOOST_AUTO_TEST_CASE(ZqMul) {
    Zq a(-684);
    Zq b(-133);
    Zq c(25435);
    BOOST_TEST(c == a * b);
    BOOST_TEST(c == b * a);
    BOOST_TEST(Zq(0) == Zq(0) * c);
    BOOST_TEST(Zq(0) == c * Zq(0));
    BOOST_TEST(c == c * Zq(1));
    BOOST_TEST(c == Zq(1) * c);
}

BOOST_AUTO_TEST_CASE(ZqSqr) {
    BOOST_TEST(Zq(1) == Zq(-1).square());
    BOOST_TEST(Zq(0) == Zq(0).square());
    BOOST_TEST(Zq(1) == Zq(1).square());
}

BOOST_AUTO_TEST_CASE(ZqSub) {
    Zq a(-1045);
    Zq b(32750);
    Zq c(31742);
    Zq d(-31742);
    BOOST_TEST(c == a - b);
    BOOST_TEST(d == b - a);
    BOOST_TEST(c == c - Zq(0));
    BOOST_TEST(Zq(0) == Zq(0) - Zq(0));
    BOOST_TEST(Zq(0) == Zq(1) - Zq(1));
}

BOOST_AUTO_TEST_CASE(ZqInv) {
    Zq a(24);
    Zq b(19115);
    Zq c(-25);
    Zq d(-5243);
    BOOST_TEST(a == b.invert().value());
    BOOST_TEST(b == a.invert().value());
    BOOST_TEST(c == d.invert().value());
    BOOST_TEST(d == c.invert().value());
    BOOST_TEST(!Zq(0).invert());
}

BOOST_AUTO_TEST_CASE(ZqInfiniteNorm) {
    Zq a(-30000);
    Zq b(30000);
    int64_t nb = 30000;
    int64_t ng = 30001;
    BOOST_TEST(!a.checkInfinityNorm(nb));
    BOOST_TEST(a.checkInfinityNorm(ng));
    BOOST_TEST(!b.checkInfinityNorm(nb));
    BOOST_TEST(b.checkInfinityNorm(ng));
}

BOOST_AUTO_TEST_SUITE_END()

File: crypto/src/test/rust/fermat.rs
/*
 * Copyright (c) 2024-2026 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use blacknet_crypto::algebra::{BalancedRepresentative, IntegerModRing, Inv, One, Square, Zero};
use blacknet_crypto::norm::InfinityNorm;

type Z = blacknet_crypto::fermat::FermatField;

#[test]
fn representative() {
    let a = Z::with_int(-1);
    let b = Z::with_int(65536);
    assert_eq!(b, a);
    assert_eq!(a.canonical(), 65536);
    assert_eq!(b.canonical(), 65536);
    assert_eq!(a.balanced(), -1);
    assert_eq!(b.balanced(), -1);
    assert_eq!(a.absolute(), 1);
    assert_eq!(b.absolute(), 1);
}

#[test]
fn add() {
    let a = Z::with_int(981);
    let b = Z::with_int(-1516);
    let c = Z::with_int(-535);
    assert_eq!(a + b, c);
    assert_eq!(b + a, c);
    assert_eq!(c + Z::ZERO, c);
    assert_eq!(Z::ZERO + c, c);
    assert_eq!(Z::ONE + Z::ZERO, Z::ONE);
    assert_eq!(Z::ZERO + Z::ONE, Z::ONE);
    assert_eq!(Z::ONE + (-Z::ONE), Z::ZERO);
}

#[test]
fn neg() {
    let a = Z::with_int(4);
    let b = Z::with_int(-4);
    assert_eq!(-a, b);
    assert_eq!(-b, a);
    assert_eq!(-Z::ZERO, Z::ZERO);
}

#[test]
fn sub() {
    let a = Z::with_int(-1045);
    let b = Z::with_int(32750);
    let c = Z::with_int(31742);
    let d = Z::with_int(-31742);
    assert_eq!(a - b, c);
    assert_eq!(b - a, d);
    assert_eq!(c - Z::ZERO, c);
    assert_eq!(Z::ZERO - Z::ZERO, Z::ZERO);
    assert_eq!(Z::ONE - Z::ONE, Z::ZERO);
}

#[test]
fn mul() {
    let a = Z::with_int(-684);
    let b = Z::with_int(-133);
    let c = Z::with_int(25435);
    assert_eq!(a * b, c);
    assert_eq!(b * a, c);
    assert_eq!(c * Z::ZERO, Z::ZERO);
    assert_eq!(Z::ZERO * c, Z::ZERO);
    assert_eq!(Z::ONE * c, c);
    assert_eq!(c * Z::ONE, c);
}

#[test]
fn sqr() {
    assert_eq!((-Z::ONE).square(), Z::ONE);
    assert_eq!(Z::ZERO.square(), Z::ZERO);
    assert_eq!(Z::ONE.square(), Z::ONE);
}

#[test]
fn inv() {
    let a = Z::with_int(24);
    let b = Z::with_int(19115);
    let c = Z::with_int(-25);
    let d = Z::with_int(-5243);
    assert_eq!(b.inv().unwrap(), a);
    assert_eq!(a.inv().unwrap(), b);
    assert_eq!(c.inv().unwrap(), d);
    assert_eq!(d.inv().unwrap(), c);
    assert!(Z::ZERO.inv().is_none());
}

#[test]
fn div() {
    let a = Z::with_int(0xE9);
    let b = Z::with_int(0x40);
    let c = Z::with_int(0x5C04);
    let d = Z::with_int(0xFEE8);
    assert_eq!((a / b).unwrap(), c);
    assert_eq!((b / a).unwrap(), d);
    assert_eq!((-Z::ONE / -Z::ONE).unwrap(), Z::ONE);
    assert!((Z::ONE / Z::ZERO).is_none());
}

#[test]
fn sum() {
    let a = [-1, -2, -3, -4, 11].map(Z::with_int);
    assert_eq!(a.into_iter().sum::<Z>(), Z::ONE);
}

#[test]
fn infinity_norm() {
    let a = Z::with_int(-30000);
    let b = Z::with_int(30000);
    let nb = 30000;
    let ng = 30001;
    assert!(!a.check_infinity_norm(&nb));
    assert!(a.check_infinity_norm(&ng));
    assert!(!b.check_infinity_norm(&nb));
    assert!(b.check_infinity_norm(&ng));
}
